### PR TITLE
feat: item like with typebox

### PR DESCRIPTION
**Pull request by Defferrard**
Hey, here a first draw about how we can implement Typebox with our architecture. Tell me what you think !

### CODE REVIEW

File: src/server.ts [hunk 1]
```diff
@@ -38,7 +39,7 @@ const start = async () => {
       },
       plugins: [ajvFormats],
     },
-  });
+  }).withTypeProvider<TypeBoxTypeProvider>();
```
Comment (spaenleh) [inline src/server.ts:42]: Should this be added to the test server as well ?

File: src/services/item/plugins/itemLike/schemas.ts [hunk 2]
```diff
@@ -1,95 +1,74 @@
...
+      // Schema Options
+      title: 'Item ID',
+      $id: 'itemId',
     },
+  ),
+);
+
+export const getLikesForMember = {
+  response: {
+    200: Type.Array(Type.Ref(packedItemLikeSchema)),
```
Comment (spaenleh) [inline src/services/item/plugins/itemLike/schemas.ts:51]:
```suggestion
    [StatusCodes.OK]: Type.Array(Type.Ref(packedItemLikeSchema)),
```

**Review by spaenleh — APPROVED** [review]
Very nice ! 

I am not sure I understand completely, about the typebox plugin: it can only be run once all schemas have been registered to it right ? but is looks like it is being executed at the very to of the app ? of is there some import magic that makes it possibe ?

**Comment by Defferrard** [comment]
> Very nice !
> 
> I am not sure I understand completely, about the typebox plugin: it can only be run once all schemas have been registered to it right ? but is looks like it is being executed at the very to of the app ? of is there some import magic that makes it possibe ?

Yes, that's it !

```typescript
export const itemLikeSchema = registerSchema(
  Type.Object(
    {
      // Object Definition
      id: Type.String({ format: 'uuid' }),
      item: Type.Ref('https://graasp.org/items/#/definitions/item'),
    },
    {
      // Schema Options
      title: 'Item Like',
      $id: 'itemLike',
    },
  ),
);
```

Since `itemLikeSchema` is initialised at the root of the file, then `registerSchema` need to be resolved to fetch its value.
It will register the schema in an array, and will be added to the `refs` when fastify register the plugin which happens on `start();` function of `server.ts` which is executed after all imports have been resolved.

**Comment by ReidyT** [comment]
> > Very nice !
> > I am not sure I understand completely, about the typebox plugin: it can only be run once all schemas have been registered to it right ? but is looks like it is being executed at the very to of the app ? of is there some import magic that makes it possibe ?
> 
> Yes, that's it !
> 
> ```ts
> export const itemLikeSchema = registerSchema(
>   Type.Object(
>     {
>       // Object Definition
>       id: Type.String({ format: 'uuid' }),
>       item: Type.Ref('https://graasp.org/items/#/definitions/item'),
>     },
>     {
>       // Schema Options
>       title: 'Item Like',
>       $id: 'itemLike',
>     },
>   ),
> );
> ```
> 
> Since `itemLikeSchema` is initialised at the root of the file, then `registerSchema` need to be resolved to fetch its value. It will register the schema in an array, and will be added to the `refs` when fastify register the plugin which happens on `start();` function of `server.ts` which is executed after all imports have been resolved.

I agree with Basile, I had to read the code twice and I finally assumed that the registred plugin will be executed after the `start()` and so the array of schema is filled at this moment. But it's a bit difficult to understand, because in the code we start by registering the plugin and then registering the schema. It is possible to register the plugin at the end? Or at least we should add some comments about that.

**Comment by Defferrard** [comment]
> > > Very nice !
> > > I am not sure I understand completely, about the typebox plugin: it can only be run once all schemas have been registered to it right ? but is looks like it is being executed at the very to of the app ? of is there some import magic that makes it possibe ?
> > 
> > 
> > Yes, that's it !
> > ```ts
> > export const itemLikeSchema = registerSchema(
> >   Type.Object(
> >     {
> >       // Object Definition
> >       id: Type.String({ format: 'uuid' }),
> >       item: Type.Ref('https://graasp.org/items/#/definitions/item'),
> >     },
> >     {
> >       // Schema Options
> >       title: 'Item Like',
> >       $id: 'itemLike',
> >     },
> >   ),
> > );
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Since `itemLikeSchema` is initialised at the root of the file, then `registerSchema` need to be resolved to fetch its value. It will register the schema in an array, and will be added to the `refs` when fastify register the plugin which happens on `start();` function of `server.ts` which is executed after all imports have been resolved.
> 
> I agree with Basile, I had to read the code twice and I finally assumed that the registred plugin will be executed after the `start()` and so the array of schema is filled at this moment. But it's a bit difficult to understand, because in the code we start by registering the plugin and then registering the schema. It is possible to register the plugin at the end? Or at least we should add some comments about that.

Here a schema in addition of my explanations
```mermaid
sequenceDiagram
    server.ts->>+schema.ts: import
    schema.ts->>typebox plugin: registerSchema(...)
    schema.ts->>typebox plugin: registerSchema(...)
    schema.ts->>typebox plugin: registerSchema(...)
    schema.ts->>-server.ts: done importing

    server.ts->>+server.ts: start()
    server.ts->>+typebox plugin: register(...)
    typebox plugin->>typebox plugin: addSchema(...)
    typebox plugin->>typebox plugin: addSchema(...)
    typebox plugin->>typebox plugin: addSchema(...)
    typebox plugin->>-server.ts: done
    deactivate server.ts
``` 

I'll add more comments to the typebox plugin

It is not possible to register the plugin at the end because Fastify need to resolve the refs.

It is a proposal to the issue that we need to `addSchema` every refs. An another solution would be to stop using refs, but I'm not quite sure it is a good idea

**Review by ReidyT — COMMENTED** [review]
Thanks for this PR, I really like the new way of creating schema, it's far less verbose! I just add a comment to maybe discuss it during the meeting before approving it.

**Review by ReidyT — APPROVED** [review]

File: src/services/item/plugins/itemLike/schemas.ts [hunk 3]
```diff
@@ -1,95 +1,74 @@
...
+import { Type } from '@sinclair/typebox';
+
+import { registerSchema } from '../../../../plugins/typebox';
+
+export const itemLikeSchema = registerSchema(
+  Type.Object(
+    {
+      // Object Definition
+      id: Type.String({ format: 'uuid' }),
+      item: Type.Ref('https://graasp.org/items/#/definitions/item'),
```
[the remark this answers is not in the document]
Comment (Defferrard) [reply src/services/item/plugins/itemLike/schemas.ts:10]: Yes it is, I'll do it step by step for the moment this ref will stay as it is, but will be change in another PR

File: src/plugins/typebox.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+/**
+ * List of schemas to be registered in the Fastify instance.
+ */
+const schemas: TSchema[] = [];
+
+/**
+ * Register a schema in the Fastify instance.
+ * This function needs be called before the server starts, so the schema will be registered in the Fastify instance.
+ * Each schema needs have defined the `$id` property.
```
[the remark this answers is not in the document]
Comment (Defferrard) [reply src/plugins/typebox.ts:13]: Typebox will throw an exception

I would like to typecheck, but I feel like it is not possible...

**Pull request merged**